app/assets/javascripts/float-kit/addon/components/d-default-toast.gjs

```diff
@@ -1,9 +1,26 @@
 import { concat, fn, hash } from "@ember/helper";
+import Modifier from "ember-modifier";
 import { or } from "truth-helpers";
 import DButton from "discourse/components/d-button";
 import concatClass from "discourse/helpers/concat-class";
 import icon from "discourse-common/helpers/d-icon";
 
+class ToastProgress extends Modifier {
+  start = Date.now();
+
+  modify(element, _, { duration }) {
+    const elapsed = Date.now() - this.start;
+    const progress = (elapsed / duration) * 100;
+    const countdownPercent = Math.max(100 - progress, 0);
+
+    element.style.width = `${countdownPercent}%`;
+
+    if (countdownPercent > 0) {
+      requestAnimationFrame(() => this.modify(element, _, { duration }));
+    }
+  }
```

Comment:
I think this would be better:

```suggestion
  modify(element, _, { duration }) {
    requestAnimationFrame(() => {
      const elapsed = Date.now() - this.start;
      const progress = (elapsed / duration) * 100;
      const countdownPercent = Math.max(100 - progress, 0);

      element.style.width = `${countdownPercent}%`;

      if (countdownPercent > 0) {
        this.modify(element, _, { duration });
      }
    });
  }
```

Reply:
Could we use the JS animation API for this?

Also, it would be best to use a [non-repainting](https://developer.mozilla.org/en-US/docs/Web/Performance/Animation_performance_and_frame_rate#css_property_cost) CSS property like `transform` rather than `width`. Should provide a smoother experience, especially on low-end devices.

Reply:
I looked at doing it this morning, but at first thought that would be more complicated for no real perf gain.

However, now that we want to pause/resume, using js animation will make that significantly easier given you we can play/pause and check state.

@dbattersby you will need to use scaleX and tweak the origin too for this to work, but besides this it should be relatively straightforward.

Reply:
Updated to use JS animation with `transform`. After looking at this again, I don't think we even need to calculate the percentage logic, we can just pass the duration to the animation and it handles that.